examples/nlp/mlm_with_bert.py

```diff
@@ -0,0 +1,209 @@
+
+
+"""
+## Set-up TPU Runtime
+"""
+
+
+def connect_to_TPU():
+    """Detect hardware, return appropriate distribution strategy"""
+    try:
+        # TPU detection. No parameters necessary if TPU_NAME environment variable is
+        # set: this is always the case on Kaggle.
+        tpu = tf.distribute.cluster_resolver.TPUClusterResolver()
+        print("Running on TPU ", tpu.master())
+    except ValueError:
+        tpu = None
+
+    if tpu:
+        tf.config.experimental_connect_to_cluster(tpu)
+        tf.tpu.experimental.initialize_tpu_system(tpu)
+        strategy = tf.distribute.experimental.TPUStrategy(tpu)
+    else:
+        # Default distribution strategy in Tensorflow. Works on CPU and single GPU.
+        strategy = tf.distribute.get_strategy()
+
+    global_batch_size = flags.BATCH_SIZE * strategy.num_replicas_in_sync
+
+    return tpu, strategy, global_batch_size
+
+
+tpu, strategy, global_batch_size = connect_to_TPU()
+print("REPLICAS: ", strategy.num_replicas_in_sync)
+
+"""
+## Load Data
+"""
+
+"""shell
+wget https://raw.githubusercontent.com/SrinidhiRaghavan/AI-Sentiment-Analysis-on-IMDB-Dataset/master/imdb_tr.csv
+"""
+
+data = pd.read_csv("imdb_tr.csv", encoding="ISO-8859-1")
+
+"""
+## Prepare Masked Language Dataset
+"""
+
+
+def regular_encode(texts, tokenizer, maxlen=512):
+    enc_di = tokenizer.batch_encode_plus(
+        texts,
+        return_attention_mask=False,
+        return_token_type_ids=False,
+        pad_to_max_length=True,
+        max_length=maxlen,
+        truncation=True,
+    )
+
+    return np.array(enc_di["input_ids"])
+
+
+tokenizer = AutoTokenizer.from_pretrained(flags.PRETRAINED_MODEL)
+X_data = regular_encode(data.text.values, tokenizer, maxlen=flags.MAX_LEN)
+
+
+def prepare_mlm_input_and_labels(X):
+    # 15% BERT masking
+    inp_mask = np.random.rand(*X.shape) < 0.15
+    # do not mask special tokens
+    inp_mask[X <= 2] = False
+    # set targets to -1 by default, it means ignore
+    labels = -1 * np.ones(X.shape, dtype=int)
+    # set labels for masked tokens
+    labels[inp_mask] = X[inp_mask]
+
+    # prepare input
+    X_mlm = np.copy(X)
+    # set input to [MASK] which is the last token for the 90% of tokens
+    # this means leaving 10% unchanged
+    inp_mask_2mask = inp_mask & (np.random.rand(*X.shape) < 0.90)
+    X_mlm[
+        inp_mask_2mask
+    ] = tokenizer.mask_token_id  # mask token is the last in the dict
+
+    # set 10% to a random token
+    inp_mask_2random = inp_mask_2mask & (np.random.rand(*X.shape) < 1 / 9)
+    X_mlm[inp_mask_2random] = np.random.randint(
+        3, tokenizer.mask_token_id, inp_mask_2random.sum()
+    )
+
+    return X_mlm, labels
+
+
+# use validation and test data for mlm
+X_train_mlm = np.vstack(X_data)
+# masks and labels
+X_train_mlm, y_train_mlm = prepare_mlm_input_and_labels(X_train_mlm)
+
+
+"""
+## Create MaskedLanguageModel using huggingface transformers
+"""
+
+
+def masked_sparse_categorical_crossentropy(y_true, y_pred):
```

Comment:
I think it's fine to keep it implemented like this -- thanks for investigating the `sample_weights` arg